### PR TITLE
Clean empty raw reads command (to be used once).

### PR DESCRIPTION
**Pull request by mberacochea**
This command will suppress all the studies that:
- Don't have any analyses jobs associated
- Share samples with other studies (the TPA study will share the samples)

Will also update all the Runs that linked to the suppressed studies,
updating the ena_study_accession (used to link to ENA on the website)

Stats after running this with a local copy of the DB:
```sql
select count(*), IS_PUBLIC from STUDY
group by IS_PUBLIC;
# +--------+---------+
# |count(*)|IS_PUBLIC|
# +--------+---------+
# |103     |0        |
# |4240    |1        |
# |249     |5        |
# +--------+---------+

select count(*) from RUN run
where run.ENA_STUDY_ACCESSION is null
# count: 418863
```

Stats before running the command:
```sql
select count(*), IS_PUBLIC from STUDY
group by IS_PUBLIC;
# +--------+---------+
# |count(*)|IS_PUBLIC|
# +--------+---------+
# |103     |0        |
# |4372    |1        |
# |117     |5        |
# +--------+---------+

select count(*) from RUN run
where run.ENA_STUDY_ACCESSION is null;
# count: 426679
```

### CODE REVIEW

File: emgapi/management/commands/clean_empty_studies.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from django.core.management import BaseCommand
+from django.conf import settings
+
+from emgapi.models import Study, StudySample, AnalysisJob, Run, Status, Sample
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    help = "Check for missing download files for all (or any) study."
```
Comment (SandyRogers) [inline emgapi/management/commands/clean_empty_studies.py:30]:
```suggestion
    help = "Suppress 'empty' Studies, i.e. studies with no analyses (provided the empty study's samples are also linked to another study)."
```

File: emgapi/management/commands/clean_empty_studies.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            # through the samples
+            samples = StudySample.objects.filter(study=study).values_list(
+                "sample_id", flat=True
+            )
+            sample_study_counts = StudySample.objects.filter(
+                sample_id__in=samples
+            ).aggregate(Count("study", distinct=True))
+            if sample_study_counts["study__count"] < 2:
+                # This is likely an empty study due to a partial upload
+                logger.info(f"{study} with no linked to other studies")
```
Comment (SandyRogers) [inline emgapi/management/commands/clean_empty_studies.py:50]:
```suggestion
                logger.info(f"{study} has no samples linked to other studies – not suppressing")
```

File: emgapi/management/commands/clean_empty_studies.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    help = "Check for missing download files for all (or any) study."
+
+    def handle(self, *args, **kwargs):
+        logger.info("Starting to cleanup the 'empty studies'")
+
+        studies = Study.objects.exclude(is_public=Status.SUPRESSED)
+        for study in studies:
+            # only those with no jobs
+            if AnalysisJob.objects.filter(study=study).exists():
+                continue
```
Comment (SandyRogers) [inline emgapi/management/commands/clean_empty_studies.py:39]: Can this be done with one query with an annotate? Not sure... If it isn't painfully slow then no worries anyway.

File: emgapi/migrations/0040_alter_study_is_public.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    dependencies = [
+        ('emgapi', '0039_run_ena_study_accession'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='study',
+            name='is_public',
+            field=models.IntegerField(db_column='IS_PUBLIC', default=False),
```
Comment (SandyRogers) [inline emgapi/migrations/0040_alter_study_is_public.py:16]: I assume this is lossless i.e. Django doesn't drop and recreate the column?

**Review by SandyRogers — CHANGES_REQUESTED** [review]
Nice! Thanks for sorting these. Couple of comments/questions :) 

Also, if this really is to be used only once... you can ignore my comments but maybe add a #TODO line to make sure we delete this once we're sure the problem is solved.

File: emgapi/management/commands/clean_empty_studies.py (hunk 3 above)
Comment (mberacochea) [reply emgapi/management/commands/clean_empty_studies.py:39]: It's fast enough..

**Comment by mberacochea** [comment]
> Nice! Thanks for sorting these. Couple of comments/questions :)
> 
> Also, if this really is to be used only once... you can ignore my comments but maybe add a #TODO line to make sure we delete this once we're sure the problem is solved.

I'll run the process and then delete the command.

File: emgapi/migrations/0040_alter_study_is_public.py (hunk 4 above)
Comment (mberacochea) [reply emgapi/migrations/0040_alter_study_is_public.py:16]: I'm going to revert this one... and will add a new field called status

File: emgapi/models.py [hunk 5]
```diff
@@ -851,7 +897,7 @@ def available(self, request, prefetch=False):
                 Prefetch('studies', queryset=Study.objects.available(request)),
                 Prefetch('metadata', queryset=SampleAnn.objects.all())
             ).defer(
-                'is_public',
+                'is_private',
```
Comment (SandyRogers) [inline emgapi/models.py:900]: Why is this deferred? Isn't it a simple no-overhead boolean column now?

File: emgapi/models.py [hunk 6]
```diff
@@ -1224,9 +1260,9 @@ def available(self, request=None):
         """
         query_filters = {
             "all": [
-                Q(study__is_public=1),
-                ~Q(sample__is_public=Sample.SUPPRESSED),
-                Q(run__status_id=Status.PUBLIC) | Q(assembly__status_id=Status.PUBLIC),
+                Q(study__is_private=False),
+                ~Q(sample__is_suppressed=True),
```
Comment (SandyRogers) [inline emgapi/models.py:1264]:
```suggestion
                Q(sample__is_suppressed=False),
```

File: emgapi/models.py [hunk 7]
```diff
@@ -1235,14 +1271,14 @@ def available(self, request=None):
         if request is not None and request.user.is_authenticated:
             username = request.user.username
             query_filters["authenticated"] = [
-                ~Q(sample__is_public=Sample.SUPPRESSED),
-                Q(study__submission_account_id=username, run__status_id=Status.PRIVATE)
+                ~Q(sample__is_suppressed=True),
```
Comment (SandyRogers) [inline emgapi/models.py:1274]:
```suggestion
                Q(sample__is_suppressed=False),
```

File: emgapi/views.py [hunk 8]
```diff
@@ -300,7 +300,7 @@ def top10(self, request):
             SAMPLE as sample
         WHERE node.lft BETWEEN parent.lft AND parent.rgt
             AND node.BIOME_ID = sample.BIOME_ID
-            AND sample.IS_PUBLIC = 1
+            AND NOT sample.IS_PRIVATE
```
Comment (SandyRogers) [inline emgapi/views.py:303]: I wonder whether this is even needed? I'd be surprised if a certain biome is oversampled by private studies to such an extent that it changes the top10 order.

File: emgapianns/management/commands/import_assembly.py [hunk 9]
```diff
@@ -81,29 +81,16 @@ def get_ena_db_assembly(self, accession):
...
-        return assembly
-
-    def update_assembly_status(self, accession, status):
-
-        logger.info("Updating assembly {} to {}".format(accession, status))
-        status = emg_models.Status.objects.using(self.emg_db).get(status=status)
-        defaults = sanitise_fields({
-            "status_id": status,
-        })
+        print(defaults)
```
Comment (SandyRogers) [inline emgapianns/management/commands/import_assembly.py:93]:
```suggestion
```
debugging leftover? or switch to `logging.info`

File: emgapianns/management/lib/create_or_update_study.py [hunk 10]
```diff
@@ -166,7 +166,7 @@ def _update_or_create_study_from_db_result(
         )
 
         hold_date = ena_study.hold_date
-        is_public = True if not hold_date else False
+        is_private = False if not hold_date else True
```
Comment (SandyRogers) [inline emgapianns/management/lib/create_or_update_study.py:169]:
```suggestion
        is_private = bool(hold_date)
```

File: tests/webuploader/test_taxonomy.py [hunk 11]
```diff
@@ -24,7 +24,7 @@
 from test_utils.emg_fixtures import *  # noqa
 
 
-@pytest.mark.usefixtures('mongodb')
+# @pytest.mark.usefixtures('mongodb')
```
Comment (SandyRogers) [inline tests/webuploader/test_taxonomy.py:27]:
```suggestion
```

**Review by SandyRogers — CHANGES_REQUESTED** [review]
Very nice!
Many inline suggestions but basically boils down to some nitpicks:
- `SUPRESS` -> `SUPPRESS` on the column names for future SQL-writing happiness.
- Possibly rename the abstract models to nouns (`SuppressibleModel` instead of `SuppressModel`)
- Since the statues are booleans now, can remove some double negatives in filters (`~Q(thing=False) -> Q(thing=True)`)

File: emgapi/models.py (hunk 5 above)
Comment (mberacochea) [reply emgapi/models.py:900]: Not sure TBH...

File: emgapi/models.py (hunk 6 above)
Comment (mberacochea) [reply emgapi/models.py:1264]: ah, yeah... mindless search/replace.. I'll change it

File: emgapi/views.py (hunk 8 above)
Comment (mberacochea) [reply emgapi/views.py:303]: I agree, I'm going to add a comment there. We can fix that later.

File: emgapianns/management/commands/import_assembly.py (hunk 9 above)
Comment (mberacochea) [reply emgapianns/management/commands/import_assembly.py:93]: ups

File: emgapianns/management/lib/create_or_update_study.py (hunk 10 above)
Comment (mberacochea) [reply emgapianns/management/lib/create_or_update_study.py:169]: :nail_care:

**Pull request merged**